File: app/src/main/java/com/spidroid/starry/adapters/PostMediaAdapter.java
package com.spidroid.starry.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;
import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.spidroid.starry.R;
import java.util.ArrayList;
import java.util.List;

public class PostMediaAdapter extends RecyclerView.Adapter<PostMediaAdapter.MediaViewHolder> {

  private List<String> imageUrls;
  private final ImageInteractionListener listener;

  public PostMediaAdapter(List<String> imageUrls, ImageInteractionListener listener) {
    this.imageUrls = new ArrayList<>(imageUrls);
    this.listener = listener;
  }

  public List<String> getImageUrls() {
    return new ArrayList<>(imageUrls);
  }

  @NonNull
  @Override
  public MediaViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    View view =
        LayoutInflater.from(parent.getContext())
            .inflate(R.layout.item_post_media, parent, false); // Keep original layout name
    return new MediaViewHolder(view);
  }

  @Override
  public void onBindViewHolder(@NonNull MediaViewHolder holder, int position) {
    String imageUrl = imageUrls.get(position);

    Glide.with(holder.itemView)
        .load(imageUrl)
        .apply(new RequestOptions().centerCrop().placeholder(R.drawable.ic_cover_placeholder))
        .into(holder.ivMedia);

    holder.itemView.setOnClickListener(v -> listener.onImageClicked(imageUrl, position));
  }

  @Override
  public int getItemCount() {
    return imageUrls.size();
  }

  public void updateUrls(List<String> newUrls) {
    List<String> oldList = this.imageUrls;
    this.imageUrls = new ArrayList<>(newUrls);

    DiffUtil.DiffResult result =
        DiffUtil.calculateDiff(
            new DiffUtil.Callback() {
              @Override
              public int getOldListSize() {
                return oldList.size();
              }

              @Override
              public int getNewListSize() {
                return imageUrls.size();
              }

              @Override
              public boolean areItemsTheSame(int oldPos, int newPos) {
                return oldList.get(oldPos).equals(imageUrls.get(newPos));
              }

              @Override
              public boolean areContentsTheSame(int oldPos, int newPos) {
                return oldList.get(oldPos).equals(imageUrls.get(newPos));
              }
            });
    result.dispatchUpdatesTo(this);
  }

  public void clear() {
    int size = imageUrls.size();
    imageUrls.clear();
    notifyItemRangeRemoved(0, size);
  }

  static class MediaViewHolder extends RecyclerView.ViewHolder {
    ImageView ivMedia;

    MediaViewHolder(@NonNull View itemView) {
      super(itemView);
      ivMedia = itemView.findViewById(R.id.ivMedia); // Keep original ID
    }
  }

  public interface ImageInteractionListener {
    void onImageClicked(String imageUrl, int position);
  }
}

File: app/src/main/java/com/spidroid/starry/adapters/PostMediaAdapter.kt
package com.spidroid.starry.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.ViewCompat
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.spidroid.starry.R
import com.spidroid.starry.databinding.ItemPostMediaBinding
import com.spidroid.starry.models.PostModel

class PostMediaAdapter(
    private val listener: PostInteractionListener
) : ListAdapter<String, PostMediaAdapter.MediaViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MediaViewHolder {
        val binding = ItemPostMediaBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return MediaViewHolder(binding)
    }

    override fun onBindViewHolder(holder: MediaViewHolder, position: Int) {
        val mediaUrl = getItem(position)
        if (mediaUrl != null) {
            holder.bind(mediaUrl)
        }
    }

    inner class MediaViewHolder(private val binding: ItemPostMediaBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(mediaUrl: String) {
            val isVideo = isVideoUrl(mediaUrl)
            binding.ivPlayButton.visibility = if (isVideo) View.VISIBLE else View.GONE
            binding.videoOverlay.visibility = if (isVideo) View.VISIBLE else View.GONE
            ViewCompat.setTransitionName(binding.ivMedia, mediaUrl)

            Glide.with(itemView.context)
                .load(mediaUrl)
                .apply(RequestOptions().centerCrop().placeholder(R.drawable.ic_cover_placeholder))
                .into(binding.ivMedia)

            itemView.setOnClickListener {
                if (bindingAdapterPosition != RecyclerView.NO_POSITION) {
                    if (isVideo) {
                        listener.onVideoPlayClicked(mediaUrl)
                    } else {
                        listener.onMediaClicked(currentList.toMutableList(), bindingAdapterPosition, binding.ivMedia)
                    }
                }
            }
        }
        private fun isVideoUrl(url: String): Boolean {
            return PostModel.VIDEO_EXTENSIONS.any { url.lowercase().endsWith(".$it") }
        }
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<String>() {
            override fun areItemsTheSame(oldItem: String, newItem: String): Boolean = oldItem == newItem
            override fun areContentsTheSame(oldItem: String, newItem: String): Boolean = oldItem == newItem
        }
    }
}